calico/felix/selectors.py

```diff
@@ -0,0 +1,664 @@
+
+    a in {"b", "c"}
+
+or
+
+    a == "b"
+
+The main entry point is the parse_selector() function, which converts the
+string representation of a selector into an object.
+
+Selector expressions follow this syntax, defined as a grammar in the
+_define_grammar() function:
+
+    label == "string_literal"  ->  comparison, e.g. my_label == "foo bar"
+    label != "string_literal"   ->  not equal; also matches if label is not
+                                    present
+    label in { "a", "b", "c", ... }  ->  true if the value of label X is one
+                                         of "a", "b", "c"
+    label not in { "a", "b", "c", ... }  ->  true if the value of label X is
+                                             not one of "a", "b", "c"
+    has(label_name)  -> True if that label is present
+    expr && expr  -> Short-circuit and
+    expr || expr  -> Short-circuit or
+    ( expr ) -> parens for grouping
+    all() or the empty selector -> matches all endpoints.
+
+"""
+
+from base64 import b64encode
+import hashlib
+import logging
+import operator
+from weakref import WeakValueDictionary
+
+from pyparsing import (QuotedString, Word, Forward, Suppress,
+                       StringEnd, delimitedList, ParseBaseException,
+                       ZeroOrMore, Keyword)
+
+from calico.datamodel_v1 import LABEL_CHARS
+
+_log = logging.getLogger(__name__)
+
+
+class ExprNode(object):
+    """
+    Base class for nodes in the AST of the grammar.
+
+    Note: to minimize occupancy, nodes should use the __slots__ mechanism
+    to name their fields.
+    """
+    __slots__ = []
+
+    def collect_reqd_values(self, pr_set):
+        pass
+
+    def collect_str_fragments(self, fragment_list):
+        """
+        Appends a series of strings to the fragment_list that, when
+        concatenated, form a string that parses to this expression.
+
+        This is better than implementing __str__ recursively because it
+        avoids an O(n^2) blow-up in the concatentations.
+
+        :param fragment_list: list of fragments to add our contribution to.
+        """
+        raise NotImplementedError()
+
+    def update_hash(self, h):
+        """
+        Updates the given hashlib hash object with a value that depends on
+        the canonical form of this expression.
+
+        Syntactically-equivalent expressions result in the same update
+        to the hash object and hence produce the same hash.  Syntactically
+        different expressions give different updates.
+        """
+        # Generate a sequence of fragments that add up to the canonical
+        # version of the expression.
+        fragments = []
+        self.collect_str_fragments(fragments)
+        # Update the hash.  Wrapping with 'selector<...>' prevents the hash
+        # from being extended in a way that would clash with something we can
+        # generate.  (Probably not an important concern but it doesn't hurt.)
+        h.update("selector<")
```

Comment:
It hardly matters, but 'node<' would make more sense, as this is just one node within a selector.